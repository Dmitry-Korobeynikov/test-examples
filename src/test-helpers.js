import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';
import { createStoreForTests } from './store/create';

export const renderShallow = (Component, props = {}, initialState = undefined) => {
  const store = createStoreForTests(initialState);
  const context = { store };
  const childContextTypes = { store: PropTypes.shape({}) };
  const shallowWrapper = shallow(<Component {...props} />, { context, childContextTypes });

  // it's connected component
  if (Component.WrappedComponent) {
    // any side effects?
    // if yes then just export mapStateToProps, mapDispatchToProps and mergeProps separately
    // one more option is to create fakeConnect by using jest.mock(...)
    shallowWrapper.connectedProps = shallowWrapper.instance()
      .selectDerivedProps(store.getState(), {}, store);
  }

  return shallowWrapper;
};

export const callRenderProp = (shallowWrapper, childSelector = null, name = 'children') => {
  return (...args) => {
    const innerWrapper = childSelector
      ? shallowWrapper.find(childSelector)
      : shallowWrapper;

    return innerWrapper.renderProp(name)(...args);
  };
};
