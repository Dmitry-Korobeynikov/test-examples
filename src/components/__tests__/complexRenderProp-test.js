import { renderShallow, callRenderProp } from '../../test-helpers';

import ComplexRenderPropComponent from '../complexRenderProp';

describe('ComplexRenderPropComponent', () => {
  const configValue = 123;

  it('should render correct layout by default', () => {
    const wrapper = renderShallow(ComplexRenderPropComponent);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct layout with not default loadingValue', () => {
    const wrapper = renderShallow(ComplexRenderPropComponent, { loadingValue: 1 });

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct layout inside SimpleRenderPropComponent by default', () => {
    const wrapper = renderShallow(ComplexRenderPropComponent);
    const innerWrapper = callRenderProp(wrapper, 'SimpleRenderPropComponent')({ configValue });

    expect(innerWrapper).toMatchSnapshot();
  });

  it('should render correct layout inside SimpleRenderPropComponent for loadingValue === configValue', () => {
    const wrapper = renderShallow(ComplexRenderPropComponent, { loadingValue: configValue });
    const innerWrapper = callRenderProp(wrapper, 'SimpleRenderPropComponent')({ configValue });

    expect(innerWrapper).toMatchSnapshot();
  });

  it('should render correct layout inside SimpleRenderPropComponent for loadingValue !== configValue', () => {
    const wrapper = renderShallow(ComplexRenderPropComponent, { loadingValue: 234 });
    const innerWrapper = callRenderProp(wrapper, 'SimpleRenderPropComponent')({ configValue });

    expect(innerWrapper).toMatchSnapshot();
  });
});
