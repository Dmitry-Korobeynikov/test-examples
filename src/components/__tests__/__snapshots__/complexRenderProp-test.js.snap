// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComplexRenderPropComponent should render correct layout by default 1`] = `
<SimpleComponent>
  <SimpleRenderPropComponent>
    <Component />
  </SimpleRenderPropComponent>
</SimpleComponent>
`;

exports[`ComplexRenderPropComponent should render correct layout with not default loadingValue 1`] = `
<SimpleComponent>
  <SimpleRenderPropComponent>
    <Component />
  </SimpleRenderPropComponent>
</SimpleComponent>
`;

exports[`ComplexRenderPropComponent should render correct layout inside SimpleRenderPropComponent by default 1`] = `
<SimpleWithConditionComponent
  isLoading={false}
/>
`;

exports[`ComplexRenderPropComponent should render correct layout inside SimpleRenderPropComponent for loadingValue !== configValue 1`] = `
<SimpleWithConditionComponent
  isLoading={false}
/>
`;

exports[`ComplexRenderPropComponent should render correct layout inside SimpleRenderPropComponent for loadingValue === configValue 1`] = `
<SimpleWithConditionComponent
  isLoading={true}
/>
`;
